python/cudf/dataframe/multiindex.py

```diff
@@ -0,0 +1,341 @@
+# Copyright (c) 2019, NVIDIA CORPORATION.
+
+import pandas as pd
+import numpy as np
+import warnings
+
+from collections.abc import Sequence
+from copy import copy, deepcopy
+
+from cudf.dataframe import columnops
+from cudf.comm.serialize import register_distributed_serializer
+from cudf.dataframe.index import StringIndex
+from . import Index
+
+
+class MultiIndex(Index):
+    """A multi-level or hierarchical index.
+
+    Provides N-Dimensional indexing into Series and DataFrame objects.
+
+    Properties
+    ---
+    levels: Labels for each category in the index hierarchy.
+    codes: Assignment of individual items into the categories of the hierarchy.
+    names: Name for each level
+    """
+
+    def __init__(self, levels, codes=None, labels=None, names=None):
+        self.names = names
+        column_names = []
+        if labels:
+            warnings.warn("the 'labels' keyword is deprecated, use 'codes' instead", FutureWarning)  # noqa: E501
+        if labels and not codes:
+            codes = labels
+        if isinstance(names, (Sequence,
+                              pd.core.indexes.frozen.FrozenNDArray,
+                              pd.core.indexes.frozen.FrozenList)):
+            if sum(x is None for x in names) > 1:
+                column_names = list(range(len(codes)))
+            else:
+                column_names = names
+        elif names is None:
+            column_names = list(range(len(codes)))
+        else:
+            column_names = names
+        if len(codes) == 0:
+            raise ValueError('MultiIndex codes can not be empty.')
+        import cudf
+        if not isinstance(codes, cudf.dataframe.dataframe.DataFrame) and\
+                not isinstance(codes[0], (Sequence,
+                               pd.core.indexes.frozen.FrozenNDArray)):
+            raise TypeError('Codes is not a Sequence of sequences')
+        if not isinstance(codes, cudf.dataframe.dataframe.DataFrame):
+            self.codes = cudf.dataframe.dataframe.DataFrame()
+            for idx, code in enumerate(codes):
+                self.codes.add_column(column_names[idx],
+                                      columnops.as_column(code))
+        else:
+            self.codes = codes
+        self.levels = levels
```

Comment:
What type of object is expected for `levels`?

Reply:
Levels is a sequence of sequences of some type, it works with `list` and `Series` and possibly `np.array`. Levels is just a list of categories at each level - it is possible to construct failure cases with lots and lots of levels, but I think that's an unlikely corner case. Codes on the other hand I think you saw has `len(levels*rows)` so storing and manipulating it efficiently was a goal.